pkg/chunk/cache/redis_client.go

```
@@ -0,0 +1,304 @@
+	var cancel context.CancelFunc
+	if c.timeout > 0 {
+		ctx, cancel = context.WithTimeout(ctx, c.timeout)
+		defer cancel()
+	}
+
+	pipe := c.rdb.TxPipeline()
+	for i := range keys {
+		pipe.Set(ctx, keys[i], values[i], c.expiration)
+	}
+	_, err := pipe.Exec(ctx)
+	return err
+}
+
+func (c *redisBasicClient) MGet(ctx context.Context, keys []string) ([][]byte, error) {
+	var cancel context.CancelFunc
+	if c.timeout > 0 {
+		ctx, cancel = context.WithTimeout(ctx, c.timeout)
+		defer cancel()
+	}
+
+	cmd := c.rdb.MGet(ctx, keys...)
+	if err := cmd.Err(); err != nil {
+		return nil, err
+	}
+
+	ret := make([][]byte, len(keys))
+	for i, val := range cmd.Val() {
+		if val != nil {
+			ret[i] = StringToBytes(val.(string))
+		}
+	}
+	return ret, nil
+}
+
+func (c *redisBasicClient) Close() error {
+	return c.rdb.Close()
+}
+
+type redisSentinelClient struct {
+	expiration time.Duration
+	timeout    time.Duration
+	rdb        *redis.SentinelClient
+	opt        *redis.Options
+}
+
+func newRedisSentinelClient(cfg *RedisConfig) *redisSentinelClient {
+	opt := &redis.Options{
+		Addr:        cfg.Endpoint,
+		Password:    cfg.Password.Value,
+		PoolSize:    cfg.PoolSize,
+		IdleTimeout: cfg.IdleTimeout,
+		MaxConnAge:  cfg.MaxConnAge,
+	}
+	if cfg.EnableTLS {
+		opt.TLSConfig = &tls.Config{}
+	}
+	return &redisSentinelClient{
+		expiration: cfg.Expiration,
+		timeout:    cfg.Timeout,
+		rdb:        redis.NewSentinelClient(opt),
+		opt:        opt,
+	}
+}
+
+func (c *redisSentinelClient) getMaster(ctx context.Context) (*redisBasicClient, error) {
+	masters, err := c.rdb.Masters(ctx).Result()
+	if err != nil {
+		return nil, err
+	}
+	if len(masters) == 0 {
+		return nil, ErrNoMasters
+	}
+	for _, master := range masters {
+		// expected: master = []interface {}{"name", "<master name>", "ip", "<IP>", "port", "<port>", ... }
+		if info, ok := master.([]interface{}); ok {
+			if len(info) >= 6 && info[2] == "ip" && info[4] == "port" {
```

Should we return error if the `info` doesn't match the expected structure?

---

I'm re-iterating on this feedback. If there are multiple `masters` (is it possible? not much familiar with Redis) and 1 of them has an unexpected `info`, we silently ignore it. is this expected or should return error anyway?